Pod/Core/GDGCondition.m

```diff
@@ -0,0 +1,187 @@
+//
+//  GDGCondition.m
+//  GoldDigger
+//
+//  Created by Pietro Caselani on 2/8/16.
+//
+
+#import <objc/runtime.h>
+#import <ObjectiveSugar/ObjectiveSugar.h>
+#import "GDGCondition.h"
+#import "GDGCondition_Protected.h"
+
+@implementation GDGCondition
+
++ (instancetype)builder
+{
+	return [[GDGCondition alloc] init];
+}
+
+#pragma mark - Initialization
+
+- (instancetype)init
+{
+	if (self = [super init])
+	{
+		_mutableTokens = [NSMutableArray array];
+		_mutableFields = [NSMutableDictionary dictionary];
+		_mutableArgs = [NSMutableDictionary dictionary];
+		_context = nil;
+
+		__weak typeof(self) weakSelf = self;
+
+		_field = ^GDGCondition *(id <GDGConditionField> field) {
+			[self appendField:field];
+
+			self.context = field.fullName;
+
+			return weakSelf;
+		};
+
+		_equals = ^GDGCondition *(id value) {
+			return [weakSelf appendValue:value forOperator:@"="];
+		};
+
+		_gt = ^GDGCondition *(id value) {
+			return [weakSelf appendValue:value forOperator:@">"];
+		};
+
+		_gte = ^GDGCondition *(id value) {
+			return [weakSelf appendValue:value forOperator:@">="];
+		};
+
+		_lt = ^GDGCondition *(id value) {
+			return [weakSelf appendValue:value forOperator:@"<"];
+		};
+
+		_lte = ^GDGCondition *(id value) {
+			return [weakSelf appendValue:value forOperator:@"<="];
+		};
+
+		_notEquals = ^GDGCondition *(id value) {
+			return [weakSelf appendValue:value forOperator:@"<>"];
+		};
+
+		_build = ^GDGCondition *(void (^builderHandler)(GDGCondition *)) {
+			return [weakSelf build:builderHandler];
+		};
+
+		_in = ^GDGCondition *(id arg) {
+			return [weakSelf appendValue:arg forOperator:@"IN"];
+		};
+
+		_cat = ^GDGCondition *(GDGCondition *builder) {
+			[weakSelf.mutableArgs addEntriesFromDictionary:builder.mutableArgs];
+			[weakSelf.mutableTokens addObjectsFromArray:builder->_mutableTokens];
+			return weakSelf;
+		};
+	}
+
+	return self;
+}
+
+#pragma mark - Build
+
+- (GDGCondition *)build:(void (^)(GDGCondition *))builder
+{
+	[_mutableTokens addObject:@"("];
+	builder(self);
+	[_mutableTokens addObject:@")"];
+	return self;
+}
+
+- (GDGCondition *)and
+{
+	[_mutableTokens addObject:@"AND"];
+	return self;
+}
+
+- (GDGCondition *)or
+{
+	[_mutableTokens addObject:@"OR"];
+	return self;
+}
+
+- (GDGCondition *)null
+{
+	[_mutableTokens addObject:@"NULL"];
+	return self;
+}
+
+- (GDGCondition *)notNull
+{
+	[_mutableTokens addObject:@"NOTNULL"];
+}
+
+- (GDGCondition *)appendValue:(id)value forOperator:(NSString *)operator
+{
+	if (!_context)
+		@throw [NSException exceptionWithName:@"Non Existent Context"
+		                               reason:@"[GDGCondition -appendValue:forOperator:] throws that you must provide a context before using an operator"
+		                             userInfo:nil];
+
+	[self.mutableTokens addObject:operator];
+
+	if ([value conformsToProtocol:@protocol(GDGConditionField)])
```

Comment:
NSArray it accepts, NSDate not yet. 
It is very hard to make all use cases work.